### ISSUE TITLE
[QST] How/Where Is the Schema Generated?

### ISSUE THREAD

**Issue by zanussbaum** [NVIDIA-Merlin/Transformers4Rec#271]
# ❓ Questions & Help
I'm curious on if the `schema.pb` file is generated manually or automatically? I have seen them in the demos and it's not entirely clear to me how they are created (if they are in the notebooks).
## Details

For example, how is [the schema](https://github.com/NVIDIA-Merlin/Transformers4Rec/tree/main/examples/getting-started-session-based) here generated?

**Comment by rnyak** [comment]
@swami-venkatesan  Please be sure you have the latest `main` branch of NVTabular library.

Are you using  your own dataset and your own NVTabular pipeline? if yes are you running  `workflow.transform(train_dataset).to_parquet(...)` step?  Here is a simple example to generate schema.pbtxt 

```
cat_features = ['movieId] >> nvt.ops.Categorify()
feats_item = cat_features['movieId'] >> ops.AddMetadata(tags=["item_id", "item"])
feats_user = cat_features['userId'] >> ops.AddMetadata(tags=[ "user_id", "user"])
feats_genres = cat_features['genres'] >> ops.AddMetadata(tags=["item"])

feats_target = (nvt.ColumnSelector(["rating"]) >> 
           ops.LambdaOp(lambda col: (col > 3).astype("int32")) >> 
           ops.AddMetadata(tags=["binary_classification", "target"]) >> 
           nvt.ops.Rename(name ='rating_binary')
          )
target_orig = ['rating'] >> ops.AddMetadata(tags=["regression", "target"])

workflow = nvt.Workflow(cat_features + feats_item + feats_user + feats_genres  + target_orig + feats_target )
train_dataset = nvt.Dataset([os.path.join(INPUT_DATA_DIR, "train.parquet")])
workflow.fit(train_dataset)

workflow.transform(train_dataset).to_parquet(
    output_path=os.path.join(INPUT_DATA_DIR, "train"),
    out_files_per_proc=1,
    shuffle=False,
)
```

You need to do  `workflow.transform(train_dataset).to_parquet(...)` step so that your schema will be saved in the `output_path`.

**Comment by swami-venkatesan** [comment]
Thanks @rnyak . It worked for me. It does add extra_metadata key to the features which I can remove easily. 

> @swami-venkatesan Please be sure you have the latest `main` branch of NVTabular library.
> 
> Are you using your own dataset and your own NVTabular pipeline? if yes are you running `workflow.transform(train_dataset).to_parquet(...)` step? Here is a simple example to generate schema.pbtxt
> 
> ```
> cat_features = ['movieId] >> nvt.ops.Categorify()
> feats_item = cat_features['movieId'] >> ops.AddMetadata(tags=["item_id", "item"])
> feats_user = cat_features['userId'] >> ops.AddMetadata(tags=[ "user_id", "user"])
> feats_genres = cat_features['genres'] >> ops.AddMetadata(tags=["item"])
> 
> feats_target = (nvt.ColumnSelector(["rating"]) >> 
>            ops.LambdaOp(lambda col: (col > 3).astype("int32")) >> 
>            ops.AddMetadata(tags=["binary_classification", "target"]) >> 
>            nvt.ops.Rename(name ='rating_binary')
>           )
> target_orig = ['rating'] >> ops.AddMetadata(tags=["regression", "target"])
> 
> workflow = nvt.Workflow(cat_features + feats_item + feats_user + feats_genres  + target_orig + feats_target )
> train_dataset = nvt.Dataset([os.path.join(INPUT_DATA_DIR, "train.parquet")])
> workflow.fit(train_dataset)
> 
> workflow.transform(train_dataset).to_parquet(
>     output_path=os.path.join(INPUT_DATA_DIR, "train"),
>     out_files_per_proc=1,
>     shuffle=False,
> )
> ```
> 
> You need to do `workflow.transform(train_dataset).to_parquet(...)` step so that your schema will be saved in the `output_path`.

**Comment by lightsailpro** [comment]
@sararb, Thanks for the explanation.  I am using pytorch 22.03 docker. The nvtabular workflow generated schema file seems missing an important annotation tag (tag: "item_id") for "item_id-list_seq" entry, which cause error later in the code (tr.TabularSequenceFeatures.from_schema)
 valueError: For masking a categorical_module is required including an item_id.

**Comment by lightsailpro** [comment]
Found a temporary workaround with sed:
#Fix the generated schema file by adding "item_id" tag in annotation section, 
!sed -i '0,/tag: "list"/{s/tag: "list"/tag: "item_id"\n    tag:  "list"/}' ./schema.pbtxt

**Comment by kontrabas380** [comment]
@swami-venkatesan @rnyak @lightsailpro 
I'm using the latest docker image (merlin-pytorch-training:latest). I tried to generate schema file with that what you said in this topic, but it is giving different schema (in my opinion in wrong style) even for the yoochoose dataset.

I added this code:
`workflow.transform(dataset).to_parquet(
    output_path='/workspace/data/'
)`

after that line:
`workflow.fit(dataset)`

As you suggested in previous comments.


And my generated file schema.pbtxt for the yoochoose dataset is different than schema_demo.pb and looks like that:
`feature {
  name: "session_id"
  type: INT
  int_domain {
    name: "session_id"
    max: 9249731
    is_categorical: true
  }
  annotation {
    tag: "categorical"
    extra_metadata {
      type_url: "type.googleapis.com/google.protobuf.Struct"
      value: "\n\021\n\013num_buckets\022\002\010\000\n\033\n\016freq_threshold\022\t\021\000\000\000\000\000\000\000\000\n\025\n\010max_size\022\t\021\000\000\000\000\000\000\000\000\n\030\n\013start_index\022\t\021\000\000\000\000\000\000\360?\n?\n\010cat_path\0223\0321workflow_etl/categories/unique.session_id.parquet\nG\n\017embedding_sizes\0224*2\n\030\n\013cardinality\022\t\021\000\000\000`x\244aA\n\026\n\tdimension\022\t\021\000\000\000\000\000\000\200@\n\034\n\017dtype_item_size\022\t\021\000\000\000\000\000\000P@\n\r\n\007is_list\022\002 \000\n\017\n\tis_ragged\022\002 \000"
    }
  }
}
feature {
  name: "item_id-count"
  type: INT
  int_domain {
    name: "item_id"
    max: 52741
    is_categorical: true
  }
  annotation {
    tag: "categorical"
    extra_metadata {
      type_url: "type.googleapis.com/google.protobuf.Struct"
      value: "\n\021\n\013num_buckets\022\002\010\000\n\033\n\016freq_threshold\022\t\021\000\000\000\000\000\000\000\000\n\025\n\010max_size\022\t\021\000\000\000\000\000\000\000\000\n\030\n\013start_index\022\t\021\000\000\000\000\000\000\360?\n<\n\010cat_path\0220\032.workflow_etl/categories/unique.item_id.parquet\nG\n\017embedding_sizes\0224*2\n\030\n\013cardinality\022\t\021\000\000\000\000\240\300\351@\n\026\n\tdimension\022\t\021\000\000\000\000\000\000\200@\n\034\n\017dtype_item_size\022\t\021\000\000\000\000\000\000@@\n\r\n\007is_list\022\002 \000\n\017\n\tis_ragged\022\002 \000"
    }
  }
}
feature {
  name: "item_id-list_seq"
  value_count {
  }
  type: INT
  int_domain {
    name: "item_id"
    max: 52741
    is_categorical: true
  }
  annotation {
    tag: "list"
    tag: "categorical"
    extra_metadata {
      type_url: "type.googleapis.com/google.protobuf.Struct"
      value: "\n\021\n\013num_buckets\022\002\010\000\n\033\n\016freq_threshold\022\t\021\000\000\000\000\000\000\000\000\n\025\n\010max_size\022\t\021\000\000\000\000\000\000\000\000\n\030\n\013start_index\022\t\021\000\000\000\000\000\000\360?\n<\n\010cat_path\0220\032.workflow_etl/categories/unique.item_id.parquet\nG\n\017embedding_sizes\0224*2\n\030\n\013cardinality\022\t\021\000\000\000\000\240\300\351@\n\026\n\tdimension\022\t\021\000\000\000\000\000\000\200@\n\034\n\017dtype_item_size\022\t\021\000\000\000\000\000\000P@\n\r\n\007is_list\022\002 \001\n\017\n\tis_ragged\022\002 \000"
    }
  }
}
feature {
  name: "category-list_seq"
  value_count {
  }
  type: INT
  int_domain {
    name: "category"
    max: 336
    is_categorical: true
  }
  annotation {
    tag: "list"
    tag: "categorical"
    extra_metadata {
      type_url: "type.googleapis.com/google.protobuf.Struct"
      value: "\n\021\n\013num_buckets\022\002\010\000\n\033\n\016freq_threshold\022\t\021\000\000\000\000\000\000\000\000\n\025\n\010max_size\022\t\021\000\000\000\000\000\000\000\000\n\030\n\013start_index\022\t\021\000\000\000\000\000\000\360?\n=\n\010cat_path\0221\032/workflow_etl/categories/unique.category.parquet\nG\n\017embedding_sizes\0224*2\n\030\n\013cardinality\022\t\021\000\000\000\000\000\000u@\n\026\n\tdimension\022\t\021\000\000\000\000\000\000E@\n\034\n\017dtype_item_size\022\t\021\000\000\000\000\000\000P@\n\r\n\007is_list\022\002 \001\n\017\n\tis_ragged\022\002 \000"
    }
  }
}
feature {
  name: "et_dayofweek_sin-list_seq"
  value_count {
  }
  type: FLOAT
  annotation {
    tag: "list"
    tag: "categorical"
    extra_metadata {
      type_url: "type.googleapis.com/google.protobuf.Struct"
      value: "\n\034\n\017dtype_item_size\022\t\021\000\000\000\000\000\000@@\n\r\n\007is_list\022\002 \001\n\017\n\tis_ragged\022\002 \000"
    }
  }
}
feature {
  name: "product_recency_days_log_norm-list_seq"
  value_count {
  }
  type: FLOAT
  annotation {
    tag: "list"
    tag: "categorical"
    extra_metadata {
      type_url: "type.googleapis.com/google.protobuf.Struct"
      value: "\n\034\n\017dtype_item_size\022\t\021\000\000\000\000\000\000P@\n\r\n\007is_list\022\002 \001\n\017\n\tis_ragged\022\002 \000"
    }
  }
}
feature {
  name: "day_index"
  type: INT
  annotation {
    tag: "categorical"
    extra_metadata {
      type_url: "type.googleapis.com/google.protobuf.Struct"
      value: "\n\034\n\017dtype_item_size\022\t\021\000\000\000\000\000\000P@\n\r\n\007is_list\022\002 \000\n\017\n\tis_ragged\022\002 \000"
    }
  }
}
`


Am I doing something wrong?

**Comment by rnyak** [comment]
@kontrabas380 you can see t[his example](https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/main/examples/getting-started-session-based/01-ETL-with-NVTabular.ipynb) as a reference how a schema file is generated from NVTabular. Please adapt your NVT pipeline based on this example. if you still have questions free free to ask.

**Comment by kontrabas380** [comment]
> @kontrabas380 you can see t[his example](https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/main/examples/getting-started-session-based/01-ETL-with-NVTabular.ipynb) as a reference how a schema file is generated from NVTabular. Please adapt your NVT pipeline based on this example. if you still have questions feel free to ask.

@rnyak, thank you for the answer - it works.
But I was asking about this specific schema:
https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/main/examples/end-to-end-session-based/schema_demo.pb

Is there possibility to generate something like this to make the example "end-to-end-session-based" working on custom dataset?

**Comment by kontrabas380** [comment]
@rnyak, I see the same problem in "tutorial" example:
https://github.com/NVIDIA-Merlin/Transformers4Rec/tree/main/examples/tutorial

Schema is already included, but it's not possible to generate a schema on a different dataset that will have the same format as the included one.

**Comment by rnyak** [comment]
@kontrabas380 thanks for letting us know. We will look into that.

**Comment by nicoladisabato** [comment]
Hi, is there any update? 
Running 

_workflow.fit_transform(dataset).to_parquet(os.path.join(INPUT_DATA_DIR, "processed_nvt"))_

the auto-generated schema.pbtxt does not work when I define the input block using tr.TabularSequenceFeatures.from_schema.

Thanks a lot.

**Comment by nicoladisabato** [comment]
First of all thanks for the reply! My problem was creating the schema.pb file.
I'm testing this ETL notebook:
_https://nvidia-merlin.github.io/Transformers4Rec/main/examples/tutorial/02-ETL-with-NVTabular.html_

In section 5.2 (creating the sequential features) this part `>> nvt.ops.AddMetadata(tags=[Tags.CATEGORICAL])` adds all annotation tag as categorical and I had the error `valueError: For masking a categorical_module is required including an item_id.`, like the @lightsailpro problem above. 
Running `workflow.output_schema` I can see that the annotation tag (tag: "item_id") for "product_id-list_seq" is not included in the tags.

Since there is no part where the schema.pb file is created, I didn't understand if in that tutorial it was necessary to modify it manually and it was not possible to create it automatically. 

After some tests, I solved this way:

```
from merlin.schema.tags import Tags
from nvtabular.ops import *

SESSIONS_MAX_LENGTH = 20 
MINIMUM_SESSION_LENGTH = 2

# Define Groupby Workflow
groupby_features = groupby_feats >> nvt.ops.Groupby(
    groupby_cols=["user_session"], 
    sort_cols=["event_time_ts"],
    aggs={
        'user_id': ['first'],
        'product_id': ["list", "count"],
        'category_id': ["list"], 
        'event_time_ts': ["first"],
        'event_time_dt': ["first"]
        },
    name_sep="-") 

# Select and truncate the sequential features
sequence_features_truncated = (
    groupby_features['category_id-list']
    >> nvt.ops.ListSlice(0,SESSIONS_MAX_LENGTH) 
)

sequence_features_truncated_item = (
    groupby_features['product_id-list']
    >> nvt.ops.ListSlice(0,SESSIONS_MAX_LENGTH) 
    >> TagAsItemID()
) 

# calculate session day index based on 'timestamp-first' column
day_index = ((groupby_features['event_time_dt-first'])  >> 
    nvt.ops.LambdaOp(lambda col: (col - col.min()).dt.days +1) >> 
    nvt.ops.Rename(f = lambda col: "day_index")
)

# Filter out sessions with length 1 (not valid for next-item prediction training and evaluation)
selected_features = (
    groupby_features['product_id-count', 'user_session'] + 
    sequence_features_truncated_item +
    sequence_features_truncated + 
    day_index
) 

filtered_sessions = selected_features >> nvt.ops.Filter(f=lambda df: df["product_id-count"] >= MINIMUM_SESSION_LENGTH)

seq_feats_list = filtered_sessions['product_id-list', 'category_id-list'] >> nvt.ops.ValueCount()

workflow = nvt.Workflow(filtered_sessions['user_session', 'product_id-count', 'day_index'] + seq_feats_list)
```

Now I have a schema like this https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/main/examples/tutorial/schema_tutorial.pb (in the form), but in the annotation there is a "extra_metadata" field, is it not a problem for the next sections? 
Thank you.

**Comment by rnyak** [comment]
@nicoladisabato yes `item_id` tag is required. Please check this [example](https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/main/examples/getting-started-session-based/01-ETL-with-NVTabular.ipynb) and try to add the tags accordingly. It should work as well. You can tag any customer dataset based on this example and it'd work.
